cli/src/commands/update/dataset.rs

```diff
@@ -0,0 +1,70 @@
+use anyhow::{Context, Result};
+use log::info;
+use reinfer_client::{Client, DatasetIdentifier, SourceId, SourceIdentifier, UpdateDataset};
+use structopt::StructOpt;
+
+/// Update a dataset.
+#[derive(Debug, StructOpt)]
+pub struct UpdateDatasetArgs {
+    #[structopt(name = "dataset")]
+    /// Name or id of the dataset to delete
+    dataset: DatasetIdentifier,
+
+    #[structopt(long = "title")]
+    /// Set the title of the dataset
+    title: Option<String>,
+
+    #[structopt(long = "description")]
+    /// Set the description of the dataset
+    description: Option<String>,
+
+    #[structopt(short = "s", long = "source")]
+    /// Names or ids of the sources in the dataset
+    sources: Option<Vec<SourceIdentifier>>,
+}
+
+pub fn update(client: &Client, args: &UpdateDatasetArgs) -> Result<()> {
+    let UpdateDatasetArgs {
+        ref dataset,
+        ref title,
+        ref description,
+        ref sources,
+    } = *args;
+
+    let source_ids = sources
+        .as_ref()
+        .map::<Result<Vec<SourceId>>, _>(|sources| {
+            sources
+                .iter()
+                .map(|source| Ok(client.get_source(source.clone())?.id))
+                .collect()
+        })
+        .transpose()
+        .context("Operation to get sources failed")?;
+
+    let dataset_full_name = match dataset.to_owned() {
+        DatasetIdentifier::FullName(name) => name,
+        dataset @ DatasetIdentifier::Id(_) => client
```

Comment:
never seen this use of `@`, here you're binding `dataset` to a `DatasetIdentifer` with id = `dataset.to_owned()` is that right?

because `dataset` can be either a full name (first match case) or an id (second match case)?

Reply:
Yes, pretty much spot on. The `variable @ pattern` syntax allows you to match with a pattern, but not deconstruct with it. This is important in Rust because normally deconstructing takes ownership. Wheras here we want to reuse the data, so the `@` allows us to avoid that.